File: code/cli/sifter_cli/commands/folders.py
from __future__ import annotations

from pathlib import Path
from typing import Optional

import typer
from sifter import Sifter
from sifter.errors import SifterError

from sifter_cli import config as cfg
from sifter_cli.output import auto_format, console, err_console

app = typer.Typer(help="Manage folders and upload documents")


def _client(api_url: Optional[str], api_key: Optional[str], profile: str) -> Sifter:
    url, key = cfg.resolve(api_url, api_key, profile)
    return Sifter(api_url=url, api_key=key)


@app.command("list")
def list_folders(
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
    as_json: bool = typer.Option(False, "--json/--table"),
    limit: int = typer.Option(50, "--limit"),
) -> None:
    """List all folders."""
    client = _client(api_url, api_key, profile)
    folders = client.list_folders(limit=limit)
    data = [{"id": f.id, "name": f.name, "document_count": f.document_count} for f in folders]
    auto_format(data, as_json)


@app.command("create")
def create_folder(
    name: str = typer.Option(..., "--name", "-n"),
    description: str = typer.Option("", "--description"),
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
    as_json: bool = typer.Option(False, "--json/--table"),
) -> None:
    """Create a new folder."""
    client = _client(api_url, api_key, profile)
    try:
        folder = client.create_folder(name=name, description=description)
        auto_format({"id": folder.id, "name": folder.name}, as_json)
    except SifterError as e:
        err_console.print(f"[red]Error:[/red] {e}")
        raise typer.Exit(2)


@app.command("upload")
def upload(
    folder_id: str = typer.Argument(...),
    path: Path = typer.Argument(..., exists=True),
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
    quiet: bool = typer.Option(False, "--quiet"),
) -> None:
    """Upload a file or directory to a folder."""
    client = _client(api_url, api_key, profile)
    folder = client.folder(folder_id)
    files = [path] if path.is_file() else sorted(path.rglob("*"))
    files = [f for f in files if f.is_file()]
    if not files:
        err_console.print("[yellow]No files found.[/yellow]")
        raise typer.Exit(1)
    try:
        for f in files:
            if not quiet:
                console.print(f"  Uploading [cyan]{f.name}[/cyan]…")
            folder.upload(str(f))
        console.print(f"[green]Uploaded {len(files)} file(s).[/green]")
    except SifterError as e:
        err_console.print(f"[red]Error:[/red] {e}")
        raise typer.Exit(2)


@app.command("link")
def link(
    folder_id: str = typer.Argument(...),
    sift_id: str = typer.Argument(...),
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
) -> None:
    """Link a folder to a sift (set default_folder_id)."""
    client = _client(api_url, api_key, profile)
    try:
        client.sift(sift_id).update(default_folder_id=folder_id)
        console.print(f"[green]Linked folder {folder_id} → sift {sift_id}.[/green]")
    except SifterError as e:
        err_console.print(f"[red]Error:[/red] {e}")
        raise typer.Exit(2)

File: code/cli/src/commands/folders.ts
import { readdir, stat } from "node:fs/promises";
import { join } from "node:path";
import { Command } from "commander";
import { autoFormat, ok, err } from "../output.js";
import { makeClient, type GlobalOpts } from "../client.js";

async function resolveFolderPath(folderPath: string, opts: GlobalOpts, create = false): Promise<string> {
  const params = new URLSearchParams({ path: folderPath });
  if (create) params.set("create", "true");
  const res = await fetch(`${opts.apiUrl}/api/folders/by-path?${params}`, {
    headers: { "X-API-Key": opts.apiKey },
  });
  if (!res.ok) throw new Error(await res.text());
  const data = await res.json() as { id: string };
  return data.id;
}

async function uploadToFolder(folderPath: string, inputPath: string, opts: GlobalOpts, quiet: boolean): Promise<number> {
  const folderId = await resolveFolderPath(folderPath, opts, true);

  const s = await stat(inputPath);
  const files = s.isDirectory()
    ? (await readdir(inputPath, { recursive: true }))
        .map(f => join(inputPath, f))
    : [inputPath];

  const { stat: statFn } = await import("node:fs/promises");
  const fileList: string[] = [];
  for (const f of files) {
    const fs = await statFn(f);
    if (fs.isFile()) fileList.push(f);
  }

  if (fileList.length === 0) {
    err("No files found.");
    process.exit(1);
  }

  const { readFile } = await import("node:fs/promises");
  const { basename } = await import("node:path");

  for (const filepath of fileList) {
    if (!quiet) process.stderr.write(`  Uploading ${basename(filepath)}…\n`);
    const bytes = await readFile(filepath);
    const form = new FormData();
    form.append("file", new Blob([bytes]), basename(filepath));
    const res = await fetch(`${opts.apiUrl}/api/folders/${folderId}/documents`, {
      method: "POST",
      headers: { "X-API-Key": opts.apiKey },
      body: form,
    });
    if (!res.ok) throw new Error(await res.text());
  }
  return fileList.length;
}

export function foldersCommand(globals: () => GlobalOpts): Command {
  const cmd = new Command("folders").description("Manage folders and upload documents");

  cmd
    .command("list")
    .description("List all folders")
    .option("--limit <n>", "Max results", "50")
    .option("--json", "Output as JSON")
    .action(async (opts) => {
      try {
        const folders = await makeClient(globals()).listFolders(parseInt(opts.limit));
        autoFormat(folders.map(f => ({ path: f.path, name: f.name, documents: f.document_count })), !!opts.json);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  cmd
    .command("create <folder-path>")
    .description("Create a folder by path (e.g. /invoices/2025), creating intermediate folders as needed")
    .option("--json", "Output as JSON")
    .action(async (folderPath, opts) => {
      try {
        const f = await makeClient(globals()).createFolder(folderPath);
        autoFormat({ path: f.path, name: f.name }, !!opts.json);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  cmd
    .command("upload <folder-path> <path>")
    .description("Upload a file or directory to a folder (folder created if needed)")
    .option("--quiet", "Suppress per-file output")
    .action(async (folderPath, inputPath, opts) => {
      try {
        const g = globals();
        const count = await uploadToFolder(folderPath, inputPath, g, !!opts.quiet);
        ok(`Uploaded ${count} file(s) to ${folderPath}.`);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  cmd
    .command("link <folder-path> <sift-id>")
    .description("Link a sift to a folder")
    .action(async (folderPath, siftId) => {
      try {
        const client = makeClient(globals());
        const folder = await client.getFolder(folderPath);
        const sift = await client.getSift(siftId);
        await folder.addSift(sift);
        ok(`Linked folder ${folderPath} → sift ${siftId}.`);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  return cmd;
}
